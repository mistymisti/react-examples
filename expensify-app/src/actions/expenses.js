import uuid from 'uuid';

//edit expense action generator
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

//This is an addExpense generator used in dispatch method of store with some default values as given below
export const addExpense = ({
    description = 'default',
    note = 'default',
    amount = 0,
    createdAt = 0
} = {}) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(),
        description,
        note,
        amount,
        createdAt
    }
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = 'default',
            note = 'default',
            amount = 0,
            createdAt = 0
        } = expenseData;
    };
};

//remove expense filter action generator for each type
export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    expense: {
        id
    }
});