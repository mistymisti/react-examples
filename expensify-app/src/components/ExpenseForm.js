import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
//import 'react-dates/lib/css/_datepicker.css'

//const date = new Date();

//const now = moment();

//console.log(now.format());

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            notes: props.expense ? props.expense.note: '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }))
    };
    onNotesChange = (e) => {
        const notes = e.target.value;
        this.setState(() => ({ notes }));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ( { amount } ));
        }
    };
    onDateChange = (createdAt) => {
        if(createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };
    onFocusChange = ( { focused } ) => {
        this.setState(() => ( { calendarFocused: focused }));
    };
    onSubmit = (e) => {
        e.preventDefault();
        console.log("submit event called !!");
        if(!this.state.amount || !this.state.description) {
            this.setState(() => ({ error: 'Please enter a valid amount or description'}));
        } else {
            this.setState(() => ( { error : ''}));
            this.props.onSubmit({
                description: this.state.description,
                notes: this.state.notes,
                createdAt: this.state.createdAt.valueOf(),
                amount: parseFloat(this.state.amount, 10) * 100
            });
        }
    };

    render() {
        return (
            <div>
                <span>{ this.state.error }</span>
                <form onSubmit={this.onSubmit}>
                    <input type="text" value= { this.state.description } placeholder="Description" autoFocus onChange = { this.onDescriptionChange } />
                    <input type="text" value = { this.state.amount } placeholder="Amount" onChange = { this.onAmountChange } />
                    <SingleDatePicker 
                    date = { this.state.createdAt }
                    onDateChange = { this.onDateChange }
                    focused = { this.state.calendarFocused }
                    onFocusChange = { this.onFocusChange }
                    numberOfMonths = {1}
                    isOutsideRange = {() => false}
                    />
                    <textarea placeholder="Add for any additional notes" value = { this.state.notes } onChange = { this.onNotesChange }></textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        )
    }
}

// const ExpenseForm = () => (
//     <div>
//         This is defuatl elemento
//     </div>
// );

// export default ExpenseForm;