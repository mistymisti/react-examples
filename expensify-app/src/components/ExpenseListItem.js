import React from 'react';
import { Link } from 'react-router-dom';

export const ExpenseListItem = ({ id, description, amount, createdAt }) => ( 
        <div>
            <Link to={`/edit/${id}`}><h1>{ description }</h1></Link>
            <p>{ amount }</p>
            <p>{ createdAt }</p>
        </div>
);
export default ExpenseListItem;