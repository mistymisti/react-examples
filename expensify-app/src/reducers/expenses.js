// Expense reducer which runs when a dispatch event is triggered by store

const expenseReducerDefaultState = [];
export default (state = expenseReducerDefaultState, action) => {
    switch(action.type) {
        case 'ADD_EXPENSE':
            return state.concat(action.expense);
        case 'REMOVE_EXPENSE':
            return [...state].filter(e => (e.id != action.expense.id));
            case 'EDIT_EXPENSE':
                return state.map(expense => {
                    if(action.id == expense.id)    
                        return {
                            ...expense,
                            ...action.updates   
                        };
                    else 
                        return expense;

                });
        default: 
            return state;
    };        
};