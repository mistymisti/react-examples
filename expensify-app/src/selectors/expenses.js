import moment from 'moment';

export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? moment(startDate).isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? moment(endDate).isSameOrBefore(createdAtMoment) : true;
        const textMatch = !text || (text && expense.description.toLowerCase().includes(text.toLowerCase()));
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        console.log(sortBy);
        if(sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
        }
        if(sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
        }
    });
};