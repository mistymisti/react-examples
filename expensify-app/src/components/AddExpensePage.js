import React from 'react';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';
import { connect } from 'react-redux';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        console.log(expense);
        this.props.addExpense(expense);
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <ExpenseForm 
                    onSubmit={ this.onSubmit }
                />
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        addExpense : (expense) => dispatch(addExpense(expense))
    };
};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);