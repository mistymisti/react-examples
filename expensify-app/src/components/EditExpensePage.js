import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startEditExpense, startRemoveExpenses } from '../actions/expenses';

export class EditExpensePage extends React.Component {
        onSubmit = (expense) => {
            console.log(expense);
            this.props.startEditExpense(this.props.expense.id, expense);
            this.props.history.push('/');
        };

        onClick = () => {
            this.props.startRemoveExpenses({ id : this.props.expense.id });
            this.props.history.push('/');
        };

        render() {
            return (
                <div>
                    <ExpenseForm 
                        expense = { this.props.expense }
                        onSubmit = { this.onSubmit }
                    />
                    <button onClick={ this.onClick } >Remove</button>
                </div>
            );
        };
}

const mapDispatchToProps = (dispatch, props) => {
    return {
        startEditExpense : (id, expense) => dispatch(startEditExpense(id, expense)),
        startRemoveExpenses : (data) => dispatch(startRemoveExpenses( data  ))
    }
};

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id
        })
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);