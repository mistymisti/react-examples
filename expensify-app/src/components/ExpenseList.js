import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from "./ExpenseListItem";
import selectedExpenses from "../selectors/expenses"

const ExpenseList = (props) => (
    <div>
        {
         props.expenses.map( (expense) => {
            return <ExpenseListItem key = { expense.description } { ...expense } />
         })}
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses : selectedExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);