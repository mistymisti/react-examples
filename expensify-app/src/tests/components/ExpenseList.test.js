import React from 'react';
import { ExpenseList } from '../../components/ExpenseList';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';

test('should test the expenselist case with expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
});

test('should test the expenselist case with no expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
});