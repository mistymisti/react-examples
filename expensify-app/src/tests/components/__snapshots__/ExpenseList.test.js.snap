// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test the expenselist case with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={0}
    createdAt={0}
    description="Gum"
    id="1"
    key="Gum"
    note=""
  />
  <ExpenseListItem
    amount={1000}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="Rent"
    note=""
  />
  <ExpenseListItem
    amount={109500}
    createdAt={345600000}
    description="test default"
    id="3"
    key="test default"
    note=""
  />
</div>
`;

exports[`should test the expenselist case with no expenses 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;
